### PR TITLE
add URL

### PR DESCRIPTION
**Pull request by jonathaningram**

### CODE REVIEW

**Comment by jonathaningram** [comment]
Abandoning this for now because we don't want to increase the surface area of the package. There is also the question of whether `env.URL` should return a string. Not every user will want this package to return a `*url.URL` as some call sites only require a string for a URL not an entire pointer/struct.

**Comment by jonathaningram** [comment]
Reopened this because this would be useful: 

This will help remove this kind of boilerplate:

```
url1Str   = env.String("URL_1", "...")
url2Str  = env.String("URL_2", "...")
url3Str  = env.String("URL_3", "...")

url1, err := url.Parse(*url1Str)
if err != nil {
	log.Fatalf("Could not parse URL 1: %v", err)
}
url2, err := url.Parse(*url2Str)
if err != nil {
	log.Fatalf("Could not parse URL 2: %v", err)
}
url3, err := url.Parse(*url3Str)
if err != nil {
	log.Fatalf("Could not parse URL 3: %v", err)
}
// ...
```

In my example I actually have 4 URLs, so this is a bit annoying.

File: check_test.go [hunk 1]
```diff
@@ -157,6 +157,51 @@ func TestIsDialAddr(t *testing.T) {
...
+		{"https://example.com/home", false},
+		{"https://example.com/home?a=b", false},
+
+		// Invalid
+		{"", true},
+		{" http://foo.com", true},    // invalid character in schema
+		{"ht tp://foo.com", true},    // invalid character in schema
+		{"1http://foo.com", true},    // invalid character in schema
+		{"http://a b.com/", true},    // no space in host name please
+		{"cache_object://foo", true}, // scheme cannot have _, relative path cannot have : in first segment
```
Comment (soulplant) [inline check_test.go]: nit: I would delete probably all but one or two examples of valid / invalid, as this isn't really testing our code, it's testing the URL parsing code.

File: env.go [hunk 2]
```diff
@@ -128,6 +129,40 @@ func (v *boolValue) String() string {
...
+	u, err := url.Parse(x)
+	if err != nil {
+		return err
+	}
+	*v = urlValue(*u)
+	return err
+}
+
+func (v *urlValue) String() string {
+	u := url.URL{
```
Comment (soulplant) [inline env.go]: Why is this necessary? If it's to redact sensitive fields you could use v.Redacted() here instead.

**Review by soulplant — APPROVED** [review]

File: env.go (hunk 2 above)
Comment (jonathaningram) [reply env.go]: Fixed. For the record the reason I did this originally is that calling String on v recurses. I needed to coerce v back to a url.URL before I can call String.

File: check_test.go (hunk 1 above)
Comment (jonathaningram) [reply check_test.go]: Done. I kept a few valid combinations just to make it obvious that you can and should be able to use full URLs including paths and query params.
